fabric/bin/peer.sh

```diff
@@ -63,6 +66,72 @@ handle_chaincode_install() {
     return
 }
 
+handle_chaincode_instantiate() {
+    OTHER_ARGS=()
+    while [[ $# > 0 ]]; do
+	case "$1" in
+	    -n|--name)
+		CC_NAME=$2;
+		shift; shift
+		;;
+	    *)
+		OTHER_ARGS+=( "$1" )
+		shift
+		;;
+	    esac
+    done
+
+    FPC_NAME=$(docker images | grep ${CC_NAME} | sed -n 's/.*-\(ecc_.*\)-.*-.*/\1/p')
```

Comment:
> This works but two suggestions:
> 
> * it is a bit brittle, better would be to use 'DOCKER_IMAGE_NAME' as defined in chaincode_install as search pattern?

DOCKER_NAME_NAME can only be used within `instantiate` because it needs `cc_name` and `cc_version`. `invoke` and `query` do not provide the version.

https://hyperledger-fabric.readthedocs.io/en/release-1.4/commands/peerchaincode.html

> * this search is used three time, why not define a separate function is_fpc_cc or alike so the search pattern is at a single place?

added abstracts for that comprising two functions `handle_chaincode_call` and `translate_fpc_name`

Reply:
The "DOCKER_IMAGE_NAME can only be used within instantiate" is not really true, what is true is "version is only provided by instantiate" as invoke and query use the version from latest instanstiated/upgraded.  So wouldn't the really robust way be to do in invoke/query a  ```../../fabric/bin/peer.sh chaincode list -C ${CHAN_ID} --instantiated | | awk "/Name: ${FPC_NAME}/ "'{ print \$4 }'```  to get the version?  (In fact, this could be right away the test, although doing a cross-check with DOCKER_IMAGE_NAME would make it even more robust) ...

Reply:
BTW: i guess we should also wrap chaincode upgrade?  (which seems like install from a wrapping perspective?)

Reply:
> BTW: i guess we should also wrap chaincode upgrade? (which seems like install from a wrapping perspective?)

this one is tricky as it allows to change the chaincode lang. That is, one could us `upgrade` to downgrade a private chaincode to a "normal" chaincode. Also, what happens when we upgrade a private chaincode and its _mrenclave_ changes. I have the feeling that we have to do a bit more here and it is better to not support upgrades at the moment. WDYT?

Reply:
> The "DOCKER_IMAGE_NAME can only be used within instantiate" is not really true, what is true is "version is only provided by instantiate" as invoke and query use the version from latest instanstiated/upgraded. So wouldn't the really robust way be to do in invoke/query a `../../fabric/bin/peer.sh chaincode list -C ${CHAN_ID} --instantiated | | awk "/Name: ${FPC_NAME}/ "'{ print \$4 }'` to get the version? (In fact, this could be right away the test, although doing a cross-check with DOCKER_IMAGE_NAME would make it even more robust) ...

mhhh ... `peer chaincode list` reveals that we install private chaincodes with this name prefix. So imagine the case a user forgot which under which name he/she has installed/instantiated a private chaincode. When checking with `peer chaincode list` it will show `ecc_somechaincode` and the user may use this name which will result in `ecc_ecc_somechaincode` invocation. This will lead to an error. WDYT?

Reply:
> > The "DOCKER_IMAGE_NAME can only be used within instantiate" is not really true, what is true is "version is only provided by instantiate" as invoke and query use the version from latest instanstiated/upgraded. So wouldn't the really robust way be to do in invoke/query a `../../fabric/bin/peer.sh chaincode list -C ${CHAN_ID} --instantiated | | awk "/Name: ${FPC_NAME}/ "'{ print \$4 }'` to get the version? (In fact, this could be right away the test, although doing a cross-check with DOCKER_IMAGE_NAME would make it even more robust) ...
> 
> mhhh ... `peer chaincode list` reveals that we install private chaincodes with this name prefix. So imagine the case a user forgot which under which name he/she has installed/instantiated a private chaincode. When checking with `peer chaincode list` it will show `ecc_somechaincode` and the user may use this name which will result in `ecc_ecc_somechaincode` invocation. This will lead to an error. WDYT?

Manipulating the output of `peer chaincode list` may be would be possible but also somehow overkill for the moment.

Reply:
Regarding upgrade changing FPC to non FPC-lang (or vice versa). I think this is simply something we cannot realistically support. Also not sure where it ever makes sense.  I guess what could work eventually is upgrade from one FPC language to another one (e.g., if ever we would add go or wasm or alike ...)

Regarding chaincode list, override of course would solve it.  I agree that we should be careful not to go overboard with peer.sh, though overriding chaincode list might just be a simple sed s/fpc_// ? (this of course assumes no other code ever uses fpc_ as prefix but i guess that's a reasonably safe assumption we can do for now?)